### PR TITLE
acceptance commands

### PR DESCRIPTION
**Pull request by patrick-wright**
Adds chef-acceptance commands `provision`, `test`, `destroy`, `verify`
`test` calls out to `provision`, `verify`, `destroy` and accepts kitchen-like destroy options
Adds support for running any test suite in the `.acceptance` dir

`chef-acceptance test my-test-suite` will run the `test.rb` recipe from the `my-test-suite` `acceptance-cookbook.`

### CODE REVIEW

**Comment by patrick-wright** [comment]
@jtimberman @schisamo @sersut

File: spec/cli_spec.rb [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+  before do
+    Dir.chdir(acceptance_dir)
   end
 
-  context 'with config' do
-    before do
-      ENV['ACCEPTANCE_ROOT'] = "#{Dir.pwd}/spec/support/acceptance"
+  shared_examples_for 'the correct recipe' do
+    it 'calls the correct acceptance recipe' do
+      cli.send(command, test_suite)
```
Comment (sersut) [inline spec/cli_spec.rb]: Maybe in the recipes we can have a `log` resource that prints out the recipe name which we check in stdout over here. That will also test that the right recipe is being run.

Comment (patrick-wright) [inline spec/cli_spec.rb]: For sure.  Right now this only tests that it doesn't blow up.

File: lib/chef-acceptance/cli.rb [hunk 2]
```diff
@@ -1,36 +1,17 @@
...
-      super
-      root = ENV['ACCEPTANCE_ROOT'] || Dir.pwd
-      acceptance_yaml = File.join(root, '.acceptance.yml')
-      if !File.exist?(acceptance_yaml)
-        puts ".acceptance.yml file not found in #{root}. Make sure you are running from an acceptance directory."
-        exit 1
+    ACCEPTANCE_COMMANDS.each do |command|
+      desc command, "Run #{command}"
+      define_method(command) do |test_suite|
+        cookbook_path = File.join(test_suite, '.acceptance')
```
Comment (sersut) [inline lib/chef-acceptance/cli.rb]: I think we should check if `cookbook_path` exists here and fail with something along the lines of "Could not find #{test_suite} in the current working directory. Make sure to have a test suite with ".acceptance" directory."

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -1,2 +1,18 @@
-$:<< File.join(File.dirname(__FILE__), '../lib')
-require 'chef-acceptance/cli'
+$LOAD_PATH << File.join(File.dirname(__FILE__), '../lib')
+
+RSpec.configure do |_config|
```
Comment (sersut) [inline spec/spec_helper.rb]: Does this needs to be inside `Rspec.configure()`?

File: test/fixtures/cookbooks/acceptance/test-suite/.acceptance/acceptance-cookbook/recipes/test.rb [hunk 4]
```diff
@@ -0,0 +1 @@
+log File.join(ENV['PWD'], __FILE__)
```
Comment (sersut) [inline test/fixtures/cookbooks/acceptance/test-suite/.acceptance/acceptance-cookbook/recipes/test.rb]: I was thinking we can `log 'test recipe'` here, add similar `provision.rb` and `destroy.rb` files and in the output check if "test" (or "provision" or "destroy") is being outputed. We will need to add two more test cases to cover the provision and destroy. Does that make sense?

File: spec/spec_helper.rb (hunk 3 above)
Comment (sersut) [inline spec/spec_helper.rb]: Also this is so cool :smile: :+1:

File: lib/chef-acceptance/cli.rb [hunk 5]
```diff
@@ -1,36 +1,25 @@
 require 'thor'
-require 'yaml'
+require 'chef-dk/chef_runner'
 
 class ChefAcceptance
   class Cli < Thor
+    ACCEPTANCE_COMMANDS = %w(destroy provision test)
+    ACCEPTANCE_DIR = '.acceptance'
```
Comment (sersut) [inline lib/chef-acceptance/cli.rb]: `.freeze` on these strings?

**Comment by sersut** [comment]
Also as a final comment if we can put `.travis.yml` and `Rakefile` here we can make sure our awesome specs and rubocop rules are running in Travis :smile:

File: test/fixtures/cookbooks/acceptance/test-suite/.acceptance/acceptance-cookbook/recipes/test.rb (hunk 4 above)
Comment (patrick-wright) [inline test/fixtures/cookbooks/acceptance/test-suite/.acceptance/acceptance-cookbook/recipes/test.rb]: My solution is pretty much the same thing.  Just more code to validate that the actual recipe path is being called.  Probably overkill, so I'm good with simplifying this.

Comment (sersut) [inline test/fixtures/cookbooks/acceptance/test-suite/.acceptance/acceptance-cookbook/recipes/test.rb]: :+1:

**Comment by sersut** [comment]
Also currently when I execute `chef-acceptance` in an invalid directory I get something like this:

```
/Users/serdar/.chefdk/gem/ruby/2.1.0/bundler/gems/chef-acceptance-3f085a9d2ce1/lib/chef-acceptance/cli.rb:14:in `block (2 levels) in <class:Cli>':  (RuntimeError)

Could not find current in the current working directory. Make sure to have a test suite with a '.acceptance/acceptance-cookbook' directory.
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Users/serdar/.chefdk/gem/ruby/2.1.0/bundler/gems/chef-acceptance-3f085a9d2ce1/bin/chef-acceptance:6:in `<top (required)>'
	from /Users/serdar/.chefdk/gem/ruby/2.1.0/bin/chef-acceptance:23:in `load'
	from /Users/serdar/.chefdk/gem/ruby/2.1.0/bin/chef-acceptance:23:in `<main>'
```

I think it would be better we just print something to the output and `exit 1` rather than raising an exception. Also we can change the wording of the message to be something like:

```
Could not find test suite 'current' in the current working directory '/Users/serdar/chef/mixlib-install/acceptance/current'. Make sure to have a test suite with a '.acceptance/acceptance-cookbook' directory.
```

**Comment by patrick-wright** [comment]
yeah... I have no idea why I did that

**Comment by patrick-wright** [comment]
@schisamo @sersut @jtimberman @chef/engineering-services This should be ready to roll

File: Gemfile [hunk 6]
```diff
@@ -1,3 +1,7 @@
 source 'https://rubygems.org'
 
 gemspec
+
+# TODO: Delete when available at rubygems
+gem 'chef-dk', github: 'chef/chef-dk',
+               branch: 'ee88c69cd1b6f5e47ae321b4d5af645340bf7f2d'
```
Comment (schisamo) [inline Gemfile]: I'm not going lie...this worries me. Bringing ChefDK (and thus Chef) along with all of its deps in seems dangerous. I would think we would want to make make the dependency tree as small as possible for this framework.

Comment (patrick-wright) [inline Gemfile]: Ok, I'll spend some time on that.

**Comment by patrick-wright** [comment]
@schisamo @sersut I just made a change to the test command.  `test` was renamed to `verify`.  `test` now works like kitchen (calls provision, verify, destroy).  I've also replicated the destroy options (passing, never, always)

**Comment by patrick-wright** [comment]
@schisamo @jtimberman I removed the dep on chefdk. Can I get some eyes on this?  I don't like the way I have to rely on a .chef/config.rb, but its working for now.  We can always optimize it later.  Let me know if you have ideas.

**Comment by schisamo** [comment]
:+1: I know it was a little more work but I think our future selves will thank us for removing the hard dep on ChefDK!

**Pull request merged**